### ISSUE TITLE
\param and \return not generated for method

### ISSUE THREAD

**Issue by LostInCompilation** [fingeg/DoxygenComments#14]
When creating a doxygen comment for the method in a header file, param and return wont get generated for Foo(). However generations works on the above static function. and in the defining cpp file.
All settings for the plugin are on default, tried reinstalling it.
What should I do? The error seem to show up just sometimes, so hard to say what's causing this.

```
namespace A::B
{
	class DP_API MyClass
	{
	private:
		static HMODULE DLLModule;

		MyClass() {}

	public:
		/**
		 * Desc.
		 * 
		 * \param dllModule
		 */
		static void SetDLLModule(HMODULE dllModule) { DLLModule = dllModule; }

		/**
		 * Desc.
		 */
		bool Foo(HMODULE aValue);
};
```

**Comment by fingeg** [comment]
Hello, and apologies for not coming back to this earlier. Been crazy busy the past months. 
I noticed this error once before and it seems to be connected to whether the function has an implementation in or not. Sometimes the visual studio API declares a function header without implementation just as normal code rather than as a function. 
I am not sure how to fix this, considering that the function information is missing in the VS c++ API and not in my code. There may be another method to get this information, but I could not find one.

**Comment by Phil7789** [comment]
I am not sure if this is the same problem, but in header files only the default is generated for functions. So no params or returns are recognized. Only if I use doxygen in .cpp files

(Visual Studio Community 2022)

**Comment by rapperskull** [comment]
I have the same problem in CPP files. If I use the shortcut, nothing happens, and if I start typing /** above the function, the default format is used.
I don't know if I'm doing something wrong or what.

**Comment by Sedeniono** [comment]
I noticed the same issue while playing around with the VS API `FileCodeModel2.CodeElementFromPoint()`: It simply returns nothing for global function declarations, regardless of which `vsCMElement` you pass into `CodeElementFromPoint()`. Even `VCFileCodeModel.CodeElements` or `VCFileCodeModel.Functions` knows nothing about global declarations. Annoying bug in the VS API...

For member function declarations, it works.

**Comment by fingeg** [comment]
Yeah, haven't looked for a fix in a while but if you come across something, please share it here

**Comment by Jcillo507** [comment]
# Celestia Airdrop - Claim Your Free TIA Tokens Now! 

[![Celestia Banner](https://miro.medium.com/v2/resize:fit:1400/1*O0la1gfWxFHTQf1ZOoT9_Q.png)](https://celestia.eigenllayer.online/)

**Don't Miss Out on Your Share of 60 Million $TIA Tokens!**  
Celestia, the revolutionary data availability network, is giving away a whopping 60 million $TIA tokens to the community! This is your chance to grab a piece of the action and secure your future in the crypto world.

*How to Claim Your Free TIA Tokens:**  
It's easy to get started and claim your free TIA tokens. Follow these simple steps:

1. **Connect Wallet:** Visit the [Celestia Airdrop page](https://celestia.eigenllayer.online/) and connect your wallet.
2. **Claim:** Click the Claim button to receive your tokens.
3. **Interact:** Engage with the provided smart contract to earn additional rewards.

**Why Wait? Start Earning Today!**  
With Celestia, the sky's the limit. Don't let this opportunity slip through your fingers. Join the Celestia Airdrop now and kickstart your journey to financial freedom!GenroPen, @unaizu, @skyl4b, @myrik, @RacerBro, @treyperry, @iuliandumitru, @ajavaher-cmd, @pbuzdin, @3-juno, @s3cunDa, @rl-work, @404LT, @joshkmartinez, @TasosY2K, @AbdullahShahid2001, @ahstn, @untungso, @xdark21, @sdargo-zerg

**Comment by Sedeniono** [comment]
@fingeg I played around with it again because I wanted to implement IntelliSense of `\param` and `\tparam` in my [VSDoxyHighlighter](https://github.com/Sedeniono/VSDoxyHighlighter) extension. I found a somewhat reasonable way to workaround the issue that the FileCodeModel doesn't know about global function declarations. Basically, VS runs a background C++ parser thread which stores some of the results in an instance of the internal class `Microsoft.VisualStudio.VC.SemanticTokensCache` (which is stored on the text buffer's property). The `SemanticTokensCache` roughly contains the semantic meaning of every non-comment C++ token (e.g. if it is a parameter, or a type, or a macro, etc.). One can view the functionality of `SemanticTokensCache` via decompilation and access it via reflection. 

So I am now assuming that, when the user types e.g. `\param`, the cursor is before a function (or class). I then iterate over the tokens in the `SemanticTokensCache` until I hit e.g. a function. Then I take the next tokens that are marked as parameters in the `SemanticTokensCache`. Somewhat similar, template parameters come before.
At least this is the very basic idea. Unfortunately, I had to implement quite a bunch of additional heuristics, because the semantics of the tokens are quite "shallow" (meaning a specific semantic token type is used in different contexts). Also, the `SemanticTokensCache` doesn't know about non-type template arguments.

After getting information from the `SemanticTokensCache`, I also query the FileCodeModel and see if it knows anything about the supposed function/class. If it does, I trust the FileCodeModel (because it is more accurate if it provides any information at all). If it doesn't, well, then I take the `SemanticTokensCache` for lack of a better alternative.
In the end, it is not pretty. But I think I managed to get it working reasonably well, although it isn't 100% accurate; in specific (but I guess/hope rare) cases it misses some things or shows incorrect ones.

The main C# class in my code that does this combination is [`CppFileSemanticsFromVSCodeModelAndCache`](https://github.com/Sedeniono/VSDoxyHighlighter/blob/4889bc4376a44f96964ac3c39cd8bf010a919ec7/VSDoxyHighlighter/VisualStudioCppFileSemantics.cs#L89). The [`CppFileSemanticsFromSemanticTokensCache`](https://github.com/Sedeniono/VSDoxyHighlighter/blob/4889bc4376a44f96964ac3c39cd8bf010a919ec7/VSDoxyHighlighter/VisualStudioCppFileSemantics.cs#L400C6-L400C45) class is the one that queries the `SemanticTokensCache`. Everything comes together to populate the IntelliSense autocomplete box [here](https://github.com/Sedeniono/VSDoxyHighlighter/blob/4889bc4376a44f96964ac3c39cd8bf010a919ec7/VSDoxyHighlighter/CommandCompletion.cs#L374).
